staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/controller/controller.go

```diff
@@ -0,0 +1,278 @@
+			utilruntime.HandleError(fmt.Errorf("error processing encryption config file %s: %v", d.filePath, err))
+			// add dummy item back to the queue to trigger file content processing.
+			d.queue.AddRateLimited(key)
+		}
+	}()
+
+	encryptionConfiguration, configChanged, err := d.processEncryptionConfig(ctx)
+	if err != nil {
+		return true
+	}
+	if !configChanged {
+		return true
+	}
+
+	// get healthz checks for all new KMS plugins.
+	if err = d.validateNewTransformersHealth(encryptionConfiguration.HealthChecks); err != nil {
+		return true
+	}
+
+	// update transformers.
+	// when reload=true there must always be one healthz check.
+	if len(encryptionConfiguration.HealthChecks) == 1 {
+		d.dynamicTransformers.Set(
+			encryptionConfiguration.Transformers,
+			closeTransformers,
+			encryptionConfiguration.HealthChecks[0],
+			encryptionConfiguration.KMSTimeout,
+		)
+
+		// update local copy of recent config content once update is successful.
+		d.lastLoadedEncryptionConfigHash = encryptionConfiguration.EncryptionFileContentHash
+		klog.V(2).InfoS("Loaded new kms encryption config content", "name", d.name)
+
+		updatedEffectiveConfig = true
+	} else {
+		err = fmt.Errorf("unexpected number of healthz checks: %d. Should have only one", len(encryptionConfiguration.HealthChecks))
+	}
+	return true
+}
+
+// loadEncryptionConfig processes the next set of content from the file.
+func (d *DynamicKMSEncryptionConfigContent) processEncryptionConfig(ctx context.Context) (
+	encryptionConfiguration *encryptionconfig.EncryptionConfiguration,
+	configChanged bool,
+	err error,
+) {
+	encryptionConfig, err := d.readEncryptionFileContent()
+	if err != nil {
+		return nil, false, err
+	}
+	if len(encryptionConfig) == 0 {
+		return nil, false, fmt.Errorf("encryption config file is empty %q", d.filePath)
+	}
+
+	// check if encryptionConfig is different from the current. Do nothing if they are the same.
+	if !d.hasEncryptionConfigChanged(encryptionconfig.ComputeEncryptionConfigHash(encryptionConfig)) {
+		return nil, false, nil
+	}
+
+	// this code path will only execute if reload=true. So passing true explicitly.
+	encryptionConfiguration, err = encryptionconfig.LoadEncryptionConfig(d.filePath, true, ctx.Done())
+
+	return encryptionConfiguration, true, err
+}
+
+func (d *DynamicKMSEncryptionConfigContent) readEncryptionFileContent() ([]byte, error) {
+	return os.ReadFile(d.filePath)
+}
+
+func (d *DynamicKMSEncryptionConfigContent) validateNewTransformersHealth(kmsPluginHealthzChecks []healthz.HealthChecker) error {
+	// test if new transformers has good health.
+	var healthCheckError error
+
+	for _, kmsPluginHealthzCheck := range kmsPluginHealthzChecks {
+		// create http get request to health check endpoint.
+		req, err := http.NewRequest(http.MethodGet, fmt.Sprintf("/healthz/%s", kmsPluginHealthzCheck.Name()), nil)
+		if err != nil {
+			return err
+		}
+
+		pollErr := wait.PollImmediate(100*time.Millisecond, 2*time.Minute, func() (bool, error) {
```

Comment:
2 minutes seems really long... it means if someone updates the config with a bad endpoint, our worker will block for 2 minutes per kms trying to health check that bad endpoint even if they've fixed the config to a good endpoint in the meantime

if we return an error, we'll requeue the config and retry with the latest content, right? trying this for something like ~10 seconds seems better

Reply:
updated to check for 10s or `kmsPluginCloseGracePeriod` if it's higher.